path-following.ts

```diff
@@ -84,6 +84,24 @@ namespace scene {
     //% path.defl="locationTiles"
     //% group="Tiles" weight=9
     export function followPath(sprite: Sprite, path: tiles.Location[], speed?: number) {
+        const currentLocation = locationOfSprite(sprite)
+        const nearestIdx = getNearestPathIdx(sprite, path)
+        const nearestTile = path[nearestIdx]
+        const remainingPath = nearestIdx === 0 ? path : path.filter((_, i) => i >= nearestIdx);
+        if (nearestTile.x === currentLocation.x && nearestTile.y ===  currentLocation.y) {
+            // already on the path
+            _followPath(sprite, remainingPath, speed);
+        }
+        else {
+            // off path, so path to it first
+            const pathToNearest = aStar(currentLocation, path[nearestIdx])
+            _followPath(sprite, pathToNearest, speed, () => {
+                _followPath(sprite, remainingPath, speed);
```

Comment:
For some reason this code doesn't seem to run in the callback

Reply:
Also, notice the logic here:
When a sprite is told to follow a path,
First find the nearest path tile,
If you're on it, follow the remainder of the path,
Otherwise, run a-star to get to the path and follow that
     Once there, follow the remainder of the original path

What do you all think? @jwunderl , @richard, @shakao

Reply:
It's not optimal, but it should work for most cases and it should be pretty efficient if the sprite is already on the path or very close by

Reply:
This will fail to find closest tile on something like this: https://makecode.com/_4RC2jiHE95hv (presuming a and b were still points on the path, it would try to go to b even though a is closest).

That doesn't matter in the given case (because they're on the same total path), but it should be easy to find the true closest tile - instead of aStar(currentLocation, path[nearestIdx]),
make it an a* with no heuristic (so just dijkstras) and make the end check be that the tile is any of the tiles in the given path. It will just move out in all directions until it reaches the any of the tiles in the list, and that's the closest reachable tile

Reply:
Yeah I originally  had something like that, but my thought was nearest via flight would be good enough and more efficient when you have a lot of sprites.

I'm imagining one of the main scenarios is you're moving a path slightly that a bunch of sprites are already on. Running dijkstras for each one with a O(n) end check seems like maybe too much..  Whereas the square-distance check is just O(n) once per sprite

Reply:
If you want the  sprite to take a truly optimal path to the destination, then you should get a fresh a-star from its position anyway I think. Because even running dijkstras to get to the nearest path tile isn't as optimal as pathing yourself to the end directly.

Reply:
But maybe dijkstra's would be fast enough it doesn't matter

Reply:
But you still find a* to path to the tile that is nearest by square distance, and in a case where there's a new wall blocking (e.g. tower defense where the wall becomes blocked and they have to back out) it'll be significantly longer than the original search to nearest tile (in the above case probably like 3 times as long?)

Reply:
You can see the inefficient pathing you pointed out here too:
![2020-05-14 20 27 29](https://user-images.githubusercontent.com/6453828/82008404-75a42400-9621-11ea-8418-9725108bdcf8.gif)

Reply:
Yeah, I would suspect it's actually significantly faster in a case like that to do the any path dijkstras than the a* to a specific one -- it does the search anyway, and in such a small area the cost between dijkstras and a* will be insignificant / you won't get the same degenerate cases where it does a search over the full map

Reply:
e.g. this case: https://makecode.com/_M79cXhbWtcDV